### PR TITLE
[Playgrounds] Add a new `-playground-option` flag to control which transforms to apply

### PR DESCRIPTION
**Pull request by abertelrud**
Generalize the existing `-playground-high-performance` flag into a set of options that control various aspects of the "playground transformation" in Sema.  The intent is for this to be a scalable way to control a more fine-grained set of upcoming options.

### Discussion

This commit adds the first two of those controllable parts of the transform, matching what the existing flag already controls (scope entry/exit and function arguments), but in an extensible way.

So instead of a single flag, we represent the playground transform options as a set of well-defined choices, with a new `-playground-option` flag to individually enable or disable those options (when prefixed with "No", the corresponding option is instead disabled). Enabling an already-enabled option or disabling an already-disabled option is a no-op.

For compatibility, the existing `-playground-high-performance` flag causes "expensive" transforms to be disabled, as before. We can also leave it as a useful shorthand to include or exclude new options even in the future, based on their cost.

The machinery for implementing this is similar to how `Features.def` works, with a new `PlaygroundOptions.def` that defines the supported playground transform options.  Each playground definition specifies the name and description, as well as whether the option is enabled by default, and whether it's also enabled in the "high performance" case.

Adding a new option in the future only requires adding it to `PlaygroundOptions.def`, deciding whether it should be on or off by default, deciding whether it should also be on or off in `-playground-high-performance` mode, and checking for its presence from the appropriate places in `PlaygroundTransform.cpp`.

Note that this is intended to control the types of user-visible results that the invoker of the compiler wants, from an externally detectable standpoint. Other flags, such as whether or not to use the extended form of the callbacks, remain as experimental features, since those deal with the mechanics and not the desired observed behavior.

### Changes

- add a `PlaygroundOption.h` that allows individual playground transform options to be defined
- add a `PlaygroundOptions.def` that defines the available playground transform options
- replace the "high performance mode" flag in `LangOptions` with a `PlaygroundOptionSet`
- add a `-playground-option` flag to `FrontendOptions.td`
- modify `CompilerInvocation.cpp` to apply the playground options to the set
   - the default value of each option, and whether "high performance mode affects it", is determined by its definition
- modify `PlaygroundTransform.cpp` to check for the new options in the set
- add a new unit test that checks that `-playground-option` transforms can be enabled
- extend the existing unit test for "high performance" mode to check that this can also be achieved by disabled corresponding options

rdar://109911673

### CODE REVIEW

**Comment by abertelrud** [comment]
@swift-ci please test

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
Test failure seems to have been due to some authorization token (i.e. infrastructure failure).  Trying again.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
> Test failure seems to have been due to some authorization token (i.e. infrastructure failure). Trying again.

That was just the full test.  The smoke test seems to be because LLDB calls into one of the methods I've changed, so I'll need to put in a bridging function to avoid revlock.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
Added back the older variant of the member function for compatibility with LLDB and any other callers.

File: include/swift/Subsystems.h [hunk 1]
```diff
@@ -131,13 +131,21 @@ namespace swift {
...
+  void performPlaygroundTransform(SourceFile &SF, PlaygroundOptionSet Opts);
+
+  /// Once type checking is complete, this optionally transforms the ASTs to
+  /// insert calls to external logging functions. This function is provided
+  /// for backward compatibility with existing code; for new code, the variant
+  /// that takes an `PlaygroundOptionSet` parameter should be used.
   ///
   /// \param HighPerformance True if the playground transform should omit
   /// instrumentation that has a high runtime performance impact.
   void performPlaygroundTransform(SourceFile &SF, bool HighPerformance);
```
Comment (artemcm) [inline include/swift/Subsystems.h:147]: Worth a comment to call out this method as deprecated and to be removed in the near future.

File: test/PlaygroundTransform/playground_options.swift [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+//
+// RUN: %empty-directory(%t)
+// RUN: cp %s %t/main.swift
+// RUN: %target-build-swift -whole-module-optimization -module-name PlaygroundSupport -emit-module-path %t/PlaygroundSupport.swiftmodule -parse-as-library -c -o %t/PlaygroundSupport.o %S/Inputs/SilentPCMacroRuntime.swift %S/Inputs/PlaygroundsRuntime.swift
+//
+// RUN: %target-build-swift -Xfrontend -playground -Xfrontend -playground-high-performance -Xfrontend -playground-option -Xfrontend FunctionParameters -Xfrontend -playground-option -Xfrontend ScopeEvents -Xfrontend -playground-option -Xfrontend ThisOptionShouldBeGracefullyIgnored -o %t/main -I=%t %t/PlaygroundSupport.o %t/main.swift
+// RUN: %target-codesign %t/main
+// RUN: %target-run %t/main | %FileCheck %s
+//
+// RUN: %target-build-swift -Xfrontend -playground -Xfrontend -playground-high-performance -Xfrontend -playground-option -Xfrontend FunctionParameters -Xfrontend -playground-option -Xfrontend ScopeEvents -Xfrontend -playground-option -Xfrontend ThisOptionShouldBeGracefullyIgnored -o %t/main2 -I=%t %t/PlaygroundSupport.o %t/main.swift
```
Comment (artemcm) [inline test/PlaygroundTransform/playground_options.swift:15]: Will we want to upgrade `-playground-option` to an official driver flag?

**Review by artemcm — APPROVED** [review]
This looks great!

File: include/swift/Subsystems.h (hunk 1 above)
Comment (abertelrud) [reply include/swift/Subsystems.h:147]: Good point — I'll add that as a comment.

File: test/PlaygroundTransform/playground_options.swift (hunk 2 above)
Comment (abertelrud) [reply test/PlaygroundTransform/playground_options.swift:15]: I'm not sure — I would think so.  Would that be a separate PR on SwiftDriver?

Comment (abertelrud) [reply test/PlaygroundTransform/playground_options.swift:15]: Having this be a driver option does sound more ergonomic.  One of the future changes I have lined up to propose as a PR is to have a way to print the available playground transform options, and that definitely feels as if it belongs in the driver (with the data coming from the frontend of course).

Comment (artemcm) [reply test/PlaygroundTransform/playground_options.swift:15]: We would need to move the option to `Options.td` and then yeah have a separate PR against swift-driver following these instructions:
https://github.com/apple/swift-driver#rebuilding-optionsswift

Comment (abertelrud) [reply test/PlaygroundTransform/playground_options.swift:15]: Thank you!  I'll take a look at that.

Comment (abertelrud) [reply test/PlaygroundTransform/playground_options.swift:15]: I've taken a closer look here, it looks to me as if all the other playground options are currently just frontend flags (including the `-playground` option itself that enables the transform).  So it looks to me as if enabling these are driver flags should be a separate PR, which would enable `-playground` as well as `-playground-option` (and ideally also a new `-playground-print-options` that would list the supported options.  @artemcm WDYT?

File: include/swift/Subsystems.h (hunk 1 above)
Comment (abertelrud) [reply include/swift/Subsystems.h:152]: Updated the PR with a comment.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Pull request merged**